### PR TITLE
Feature/61 configuration validation

### PR DESCRIPTION
**Pull request by tomblench**
**Feature PR for enhanced configuration and validation**

- Breaking changes to sink and source options (see README and CHANGELOG)
- Support for all auth methods

See #61, #67, #68

### CODE REVIEW

**Review by ricellis — CHANGES_REQUESTED** [review]
A few suggestions, but on the whole I think this is good.

File: README.md [hunk 1]
```diff
@@ -67,9 +70,12 @@ Parameter | Value | Required | Default value | Description
...
 topics|\<topic1\>,\<topic2\>,..|YES|None|The list of topics you want to consume messages from.
-cloudant.db.url|https://\<account\>.cloudant.com/\<database\>|YES|None|The Cloudant database to write documents to.
-cloudant.db.username|\<username\>|YES|None|The Cloudant username to use for authentication.
-cloudant.db.password|\<password\>|YES|None|The Cloudant password to use for authentication.
+cloudant.url|https://\<your-account\>.cloudant.com|YES|None|The Cloudant server to write documents to.
+cloudant.db|\<your-db\>|YES|None|The Cloudant database to write documents to.
+cloudant.auth.type|`iam`\|`couchdb_session`\|`basic`\|`noauth`|NO|`iam`|The Cloudant authentication type to use (case insensitive).
+cloudant.apikey|\<your-apikey\>|NO|None|The Cloudant IAM API key to use for authentication (when using `iam` authentication).
+cloudant.username|\<your-username\>|NO|None|The Cloudant username to use for authentication (when using `couchdb_session` or `basic` authentication).
+cloudant.password|\<your-password\>|NO|None|The Cloudant password to use for authentication (when using `couchdb_session` or `basic` authentication).
```
[the remark this answers is not in the document]
Comment (tomblench) [reply README.md]: See updated README with dedicated section for auth options

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantConnectorConfig.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                // Kafka topic
+                .define(InterfaceConst.TOPIC,
+                        Type.LIST,
+                        Importance.HIGH,
+                        ResourceBundleUtil.get(MessageKey.KAFKA_TOPIC_LIST_DOC),
+                        DATABASE_GROUP,
+                        1,
+                        Width.LONG,
+                        ResourceBundleUtil.get(MessageKey.KAFKA_TOPIC_LIST_DISP));
+    }
```
[the remark this answers is not in the document]
Comment (tomblench) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantConnectorConfig.java:214]: Done - for now all at `Importance.HIGH` and in `DATABASE_GROUP` - not sure if that's right but I don't think it's a blocker

File: src/main/java/com/ibm/cloudant/kafka/connect/ListRecommender.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        if (value == null) {
+            return;
+        }
+
+        if (value instanceof String) {
+            value = ((String)value).toLowerCase();
+        }
+
+        if (!validValues.contains(value)) {
+            throw new ConfigException(key, value, "Value must be one of: " + this.validValues);
```
[the remark this answers is not in the document]
Comment (tomblench) [reply src/main/java/com/ibm/cloudant/kafka/connect/ListRecommender.java]: Done - added to `MessageKey` and `messages_en_US.properties`

File: src/main/java/com/ibm/cloudant/kafka/connect/UrlValidator.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        if (value instanceof String) {
+            String stringValue = (String)value;
+            try {
+                new URL(stringValue);
+            } catch (MalformedURLException mue) {
+                throw new ConfigException(name, value, "Value not a URL: " + value);
+            }
+            return;
+        }
+        throw new ConfigException(name, value, "Value not a URL: " + value);
```
[the remark this answers is not in the document]
Comment (tomblench) [reply src/main/java/com/ibm/cloudant/kafka/connect/UrlValidator.java]: Done

File: README.md [hunk 5]
```diff
@@ -45,10 +45,13 @@ Parameter | Value | Required | Default value | Description
 name|cloudant-source|YES|None|A unique name to identify the connector with.
 connector.class|com.ibm.cloudant.kafka.connect.CloudantSourceConnector|YES|None|The connector class name.
 topics|\<topic1\>,\<topic2\>,..|YES|None|A list of topics you want messages to be written to.
-cloudant.db.url|https://\<account\>.cloudant.com/\<database\>|YES|None|The Cloudant database to read documents from.
-cloudant.db.username|\<username\>|YES|None|The Cloudant username to use for authentication.
-cloudant.db.password|\<password\>|YES|None|The Cloudant password to use for authentication.
-cloudant.db.since|1-g1AAAAETeJzLYWBgYMlgTmGQT0lKzi9..|NO|0|The first change sequence to process from the Cloudant database above. 0 will apply all available document changes.
+cloudant.url|https://\<your-account\>.cloudant.com|YES|None|The Cloudant server to read documents from.
```
[the remark this answers is not in the document]
Comment (ricellis) [reply README.md]: You missed this in the table, but did add it in the example block so I'm fine, a URL is a URL 😄

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantConnectorConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ricellis) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantConnectorConfig.java:214]: Agree I honestly find the H, M, L importance a bit hard to fathom, it really depends on the requirements. Clearly some properties are of very high importance, but only when using specific auth types or features. I think going with this for now is OK.

**Review by ricellis — APPROVED** [review]

**Review by emlaver — APPROVED** [review]

**Pull request merged**